nearcore/src/cold_storage.rs

```diff
@@ -224,65 +224,88 @@ enum ColdStoreInitialMigrationResult {
+///   This process will take a long time. Cold head will be set to hot final head BEFORE the migration started.
+/// - If hot storage is of type `Hot`, this node was just created with split storage,
+///   and we need to perform migration of genesis data to cold storage.
+///   Cold head will be set to genesis height.
+/// - Other kinds of hot storage are indicative of configuration error.
+///
+/// After cold head is determined this function
+/// 1. performs migration
+/// 2. updates cold head
 ///
 /// Any Ok status means that this function should not be retried:
 /// - either migration was performed (now or earlier)
 /// - or migration was interrupted, which means the `keep_going` flag was set to `false`
 ///   which means that everything cold store thread related has to stop
 ///
 /// Error status means that for some reason migration cannot be performed.
-fn cold_store_initial_migration(
+fn cold_store_migration(
     split_storage_config: &SplitStorageConfig,
     keep_going: &Arc<AtomicBool>,
+    genesis_height: BlockHeight,
     hot_store: &Store,
     cold_store: &Store,
     cold_db: &Arc<ColdDB>,
-) -> anyhow::Result<ColdStoreInitialMigrationResult> {
-    // We only need to perform the migration if hot store is of kind Archive and cold store doesn't have a head yet
-    if hot_store.get_db_kind()? != Some(near_store::metadata::DbKind::Archive)
-        || cold_store.get(DBCol::BlockMisc, HEAD_KEY)?.is_some()
-    {
-        return Ok(ColdStoreInitialMigrationResult::NoNeedForMigration);
+) -> anyhow::Result<ColdStoreMigrationResult> {
+    // Migration is only needed if cold storage is not properly initialised,
+    // i.e. if cold head is not set.
+    if cold_store.get(DBCol::BlockMisc, HEAD_KEY)?.is_some() {
+        return Ok(ColdStoreMigrationResult::NoNeedForMigration);
     }
 
-    tracing::info!(target: "cold_store", "Starting initial population of cold store");
+    tracing::info!(target: "cold_store", "Starting population of cold store.");
+    let new_cold_height = match hot_store.get_db_kind()? {
+        None => {
+            tracing::error!(target: "cold_store", "Hot store DBKind not set.");
+            return Err(anyhow::anyhow!("Hot store DBKind is not set"));
+        }
+        Some(near_store::metadata::DbKind::Hot) => {
+            tracing::info!(target: "cold_store", "Hot store DBKind is Hot.");
+            genesis_height
+        }
+        Some(near_store::metadata::DbKind::Archive) => {
+            tracing::info!(target: "cold_store", "Hot store DBKind is Archive.");
+            hot_store
+                .get_ser::<Tip>(DBCol::BlockMisc, FINAL_HEAD_KEY)?
+                .ok_or_else(|| anyhow::anyhow!("FINAL_HEAD not found in hot storage"))?
+                .height
+        }
+        Some(kind) => {
+            tracing::error!(target: "cold_store", "Hot store DBKind not {kind:?}.");
+            return Err(anyhow::anyhow!(format!("Hot store DBKind not {kind:?}.")));
```

Comment:
This error message is a bit confusing. How about:
"Hot store DbKind not supported. kind={kind:?}."
nit: In tracing prefer to use the structured logs instead of formatted strings:
```
tracing::error!(target: "cold_store", ?kind, "Hot store DbKind not supported.");
```